### PR TITLE
fix: disable checking for void return

### LINKED ISSUE #1197: docs: add note about how to resolve `no-misused-promises` in JSX attributes
### Provide environment information

```
System:
    OS: Windows 10 10.0.19045
    CPU: (24) x64 AMD Ryzen 9 3900XT 12-Core Processor
    Memory: 8.32 GB / 15.92 GB
  Binaries:
    Node: 19.2.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 9.4.0 - C:\Program Files\nodejs\npm.CMD
    
"ct3aMetadata": {
    "initVersion": "7.5.1"
  }
```

### Describe the bug

TypeScript ESLint throws this error: "Promise-returning function provided to attribute where a void return was expected." in React Hook Forms handleSubmit function. 

I can't change the library code to return a Promise rather than void, and I believe that React Hook Form's TypeScript configuration they used when building the library was less strict than t3's. 

I'm not alone in this issue, as others in the discord seem to have encountered it with React Hook Forms and likely with other libraries. According to official TS ESLint documentation it is recommended to change the .eslintrc.json config file to include this: 

```json
"rules": {
  "@typescript-eslint/no-misused-promises": [2, {
    "checksVoidReturn": {
      "attributes": false
    }
  }]
},
```

I'll open a PR soon

### Reproduction repo

https://github.com/esotuvaka/change-tracker/blob/main/src/pages/issues/index.tsx

### To reproduce

In this file I'm using react-hook-form, and have the default tsconfig settings that come with create-t3-app. Setting onSubmit={onSubmit} shows the error: 

"Promise-returning function provided to attribute where a void return was expected.eslint[@typescript-eslint/no-misused-promises](https://typescript-eslint.io/rules/no-misused-promises)"

In general, to recreate use the current default tsconfig settings and try to feed a Promise-returning function to one that normally returns void. 

```ts
import { useForm } from "react-hook-form";
import type { Resolver } from "react-hook-form";

type FormValues = {
  title: string;
  description: string;
  environment: string;
};

const resolver: Resolver<FormValues> = async (values) => {
  return Promise.resolve({
    values: !values.title ? {} : values,
    errors: !values.title
      ? {
          title: {
            type: "required",
            message: "This is required.",
          },
        }
      : {},
  });
};

export default function Issues() {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>({
    resolver: resolver,
  });
  const onSubmit = handleSubmit((data) => alert(JSON.stringify(data)));

  return (
    <section className="flex min-h-screen flex-col items-center justify-center bg-gradient-to-b from-neutral-700 to-black pt-[10vh]">
      <div className="mx-auto h-[90vh] w-4/5 text-5xl text-white">
        <div>
          <h1>Add new issue</h1>
          <form className="flex flex-col" onSubmit={onSubmit}>
            <label>Title</label>
            <input {...register("title")} />
            {errors.title ? <p>{errors.title.message}</p> : <></>}
            <label>Description</label>
            <input {...register("description")} />
            <label>Environment</label>
            <input
              {...register("environment")}
              placeholder="Steps to reproduce the issue"
            />
            <input type="submit" />
          </form>
        </div>
      </div>
    </section>
  );
}
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by esotuvaka**
Closes #1197

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

## Changelog

Changed www's .eslintrc.json to stop checking functions for void returns. This allows external libraries like react-hook-form to be implemented, as their documentation that recommends feeding a return-promising function to one that expects to return void. This was despite react-hook-form being built in TypeScript.

Disabling this check will allow other library code using less strict versions of TypeScript to be implemented without errors, and is the official TS ESLint solution: 

https://github.com/typescript-eslint/typescript-eslint/pull/4623

https://github.com/react-hook-form/react-hook-form/discussions/9325#discussioncomment-4060566

---

### CODE REVIEW

**Comment by esotuvaka** [comment]
just realized that my issue is with the .eslintrc.json file that you get when you create-t3-app, not the .eslintrc.cjs that is present in the repo that scaffolds t3 apps. Not sure how to change what the .eslintrc.json rules are for people that create a t3 app, but this pull request is irrelevant to the actual issue #1197

File: cli/template/base/.eslintrc.json [hunk 1]
```diff
@@ -17,6 +17,15 @@
...
-    "@typescript-eslint/consistent-type-imports": "warn"
+    "@typescript-eslint/consistent-type-imports": "warn",
+    "@typescript-eslint/no-misused-promises": [
+          2,
+          {
+            "checksVoidReturn": {
+              "attributes": false
+            }
+          }
+        ]
```
Comment (juliusmarminge) [inline cli/template/base/.eslintrc.json:28]:
```suggestion
    "@typescript-eslint/no-misused-promises": [
      "error",
      {
        "checksVoidReturn": {
          // Allow `void` return type for JSX attributes such as `onClick` / `onSubmit`
          "attributes": false
        }
      }
    ]
```

File: www/.eslintrc.cjs [hunk 2]
```diff
@@ -39,7 +39,7 @@ module.exports = {
         "no-unused-vars": [
           "error",
           { argsIgnorePattern: "^_", destructuredArrayIgnorePattern: "^_" },
-        ],
+        ],
```
Comment (juliusmarminge) [inline www/.eslintrc.cjs:42]:
```suggestion
        ],
```

File: cli/template/base/.eslintrc.json [hunk 3]
```diff
@@ -17,6 +17,16 @@
...
+      "error",
+      {
+        "checksVoidReturn": {
+          // Allow `void` return type for JSX attributes such as `onClick` / `onSubmit`
+          "attributes": false
+        }
+      }
+    ]
   }
+
```
Comment (juliusmarminge) [inline cli/template/base/.eslintrc.json:31]:
```suggestion
```

**Comment by esotuvaka** [comment]
Might be able to make this even more specific to just this conflicting return types issue by using: 

"@typescript-eslint/no-misused-promises": [
    "error",
    {
      "checksVoidReturn": {
        "returns": false
      }
    }
  ]

  "For example, if you don't mind that passing a () => Promise<void> to a () => void parameter or JSX attribute can lead to a 
  floating unhandled Promise." Might need a comment mentioning this, so any imported library code needs to explicitly handle 
  unresolved Promise errors.
  
  Quote TS ESLint docs: https://typescript-eslint.io/rules/no-misused-promises/

**Comment by juliusmarminge** [comment]
> Might be able to make this even more specific to just this conflicting return types issue by using:
> 
> ```json
> "@typescript-eslint/no-misused-promises": [
>     "error",
>     {
>       "checksVoidReturn": {
>         "returns": false
>       }
>     }
>   ]
> ```
> 
> "For example, if you don't mind that passing a () => Promise to a () => void parameter or JSX attribute can lead to a floating unhandled Promise." Might need a comment mentioning this, so any imported library code needs to explicitly handle unresolved Promise errors.
> 
> Quote TS ESLint docs: https://typescript-eslint.io/rules/no-misused-promises/

What do you think is the best approach? If eslint says this is the way we should probably follow it?

**Comment by esotuvaka** [comment]
I think a doc change. Possibly in FAQs we'd add in a section on "Type errors with external libraries", where we can link recommended ESLint workarounds / settings. 

The framework is so type strict currently that a suggested list of workarounds for small, specific stuff like Promises in a void return from a 3rd party library would be a cool feature to have, reduce stack overflow / google searching, and improve DX. 

I'll change my commit from the config change to an FAQ change once I'm home later. Thank you @juliusmarminge !

**Comment by juliusmarminge** [comment]
I'll close this but feel free to open a PR for a docs addition when you got time :)

**Pull request closed without merging**